pkg/store/cache/caching_bucket.go

```diff
@@ -132,35 +131,184 @@ func (cb *CachingBucket) ReaderWithExpectedErrs(expectedFunc objstore.IsOpFailur
+	if err == nil && remainingTTL > 0 {
+		if data, err := cfg.codec.Encode(list); err == nil {
+			cfg.cache.Store(ctx, map[string][]byte{key: data}, remainingTTL)
+		} else {
+			level.Warn(cb.logger).Log("msg", "failed to encode Iter result", "err", err)
+		}
+	}
+	return err
+}
+
+func (cb *CachingBucket) Exists(ctx context.Context, name string) (bool, error) {
+	cfgName, cfg := cb.cfg.findExistConfig(name)
+	if cfg == nil {
+		return cb.Bucket.Exists(ctx, name)
+	}
+
+	cb.operationRequests.WithLabelValues(opExists, cfgName).Inc()
+
+	key := cachingKeyExists(name)
+	hits := cfg.cache.Fetch(ctx, []string{key})
+
+	if ex := hits[key]; ex != nil {
+		exists, err := strconv.ParseBool(string(ex))
+		if err == nil {
+			cb.operationHits.WithLabelValues(opExists, cfgName).Inc()
+			return exists, nil
+		}
+		level.Warn(cb.logger).Log("msg", "unexpected cached 'exists' value", "val", string(ex))
+	}
+
+	existsTime := time.Now()
+	ok, err := cb.Bucket.Exists(ctx, name)
+	if err == nil {
+		storeExistsCacheEntry(ctx, key, ok, existsTime, cfg.cache, cfg.existsTTL, cfg.doesntExistTTL)
+	}
+
+	return ok, err
+}
+
+func storeExistsCacheEntry(ctx context.Context, cachingKey string, exists bool, ts time.Time, cache cache.Cache, existsTTL, doesntExistTTL time.Duration) {
+	var ttl time.Duration
+	if exists {
+		ttl = existsTTL - time.Since(ts)
+	} else {
+		ttl = doesntExistTTL - time.Since(ts)
+	}
+
+	if ttl > 0 {
+		cache.Store(ctx, map[string][]byte{cachingKey: []byte(strconv.FormatBool(exists))}, ttl)
+	}
+}
+
+func (cb *CachingBucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	cfgName, cfg := cb.cfg.findGetConfig(name)
+	if cfg == nil {
+		return cb.Bucket.Get(ctx, name)
+	}
+
+	cb.operationRequests.WithLabelValues(opGet, cfgName).Inc()
+
+	contentKey := cachingKeyContent(name)
+	existsKey := cachingKeyExists(name)
+
+	hits := cfg.cache.Fetch(ctx, []string{contentKey, existsKey})
+	if hits[contentKey] != nil {
+		cb.operationHits.WithLabelValues(opGet, cfgName).Inc()
+		return ioutil.NopCloser(bytes.NewReader(hits[contentKey])), nil
+	}
+
+	// If we know that file doesn't exist, we can return that. Useful for deletion marks.
+	if ex := hits[existsKey]; ex != nil {
+		if exists, err := strconv.ParseBool(string(ex)); err == nil && !exists {
+			cb.operationHits.WithLabelValues(opGet, cfgName).Inc()
+			return nil, errObjNotFound
+		}
+	}
+
+	getTime := time.Now()
+	reader, err := cb.Bucket.Get(ctx, name)
+	if err != nil {
+		if cb.Bucket.IsObjNotFoundErr(err) {
+			// Cache that object doesn't exist.
+			storeExistsCacheEntry(ctx, existsKey, false, getTime, cfg.cache, cfg.existsTTL, cfg.doesntExistTTL)
+		}
+
+		return nil, err
+	}
+	defer runutil.CloseWithLogOnErr(cb.logger, reader, "CachingBucket.Get(%q)", name)
+
+	data, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+
+	ttl := cfg.contentTTL - time.Since(getTime)
+	if ttl > 0 {
+		cfg.cache.Store(ctx, map[string][]byte{contentKey: data}, ttl)
```

Comment:
ditto with too small ttl

Also I am not sure if it makes sense to be size unbounded.....

Reply:
:thinking:

Reply:
Not anymore. PTAL.